### PR TITLE
hifi: #21216 - Resize entities with hand controllers while editing

### PR DESCRIPTION
**Pull request by CraftsmanMG**
hifi: #21216 - Resize entities with hand controllers while editing.
Modified StretchTool to allow for movement in 3D
- uses data from motion controller (position and rotation) to
stretch/resize the entities/primitives in 3D.
- the range of the resize depends on distance to the object.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

**Comment by ZappoMan** [comment]
ok to test

File: scripts/system/libraries/entitySelectionTool.js [hunk 1]
```diff
@@ -2603,18 +2614,23 @@ SelectionDisplay = (function() {
...
             z: Math.abs(direction.z) > 0 ? 1 : 0,
         };
+		
+		
 
         var numDimensions = mask.x + mask.y + mask.z;
 
         var planeNormal = null;
         var lastPick = null;
+		var lastPick3D = null;
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: something is wrong with your formatting here... coding standard says 4 space indents. Please fix.

File: scripts/system/libraries/entitySelectionTool.js [hunk 2]
```diff
@@ -2652,8 +2668,18 @@ SelectionDisplay = (function() {
 
             // Scaled offset in world coordinates
             var scaledOffsetWorld = vec3Mult(initialDimensions, offsetRP);
+			
             pickRayPosition = Vec3.sum(initialPosition, Vec3.multiplyQbyV(rotation, scaledOffsetWorld));
-
+			// Scaled offset in world coordinates for 3D manipulation
+			if (directionFor3DStretch){
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: space between `)` and `{`

File: scripts/system/libraries/entitySelectionTool.js [hunk 3]
```diff
@@ -2652,8 +2668,18 @@ SelectionDisplay = (function() {
...
+				// Scale pivot to be in the same range as registrationPoint
+				var scaledPivot3D = Vec3.multiply(0.5, Vec3.multiply(1.0, directionFor3DStretch));
+				deltaPivot3D = Vec3.subtract(centeredRP, scaledPivot3D);
+				
+				var scaledOffsetWorld3D = vec3Mult(initialDimensions, 
+				Vec3.subtract(Vec3.multiply(0.5, Vec3.multiply(-1.0, directionFor3DStretch)), centeredRP));
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: multi-line statements should have follow on lines indented

File: scripts/system/libraries/entitySelectionTool.js [hunk 4]
```diff
@@ -2754,12 +2780,25 @@ SelectionDisplay = (function() {
...
                 pickRayPosition,
                 planeNormal);
-
+				
+			var planeNormal3D = {
+                        x: 0,
+                        y: 0,
+                        z: 0
+                    };
+			if (directionFor3DStretch){
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: spaces

File: scripts/system/libraries/entitySelectionTool.js [hunk 5]
```diff
@@ -2754,12 +2780,25 @@ SelectionDisplay = (function() {
...
+				lastPick3D = rayPlaneIntersection(pickRay,
+                pickRayPosition3D,
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: indents

File: scripts/system/libraries/entitySelectionTool.js [hunk 6]
```diff
@@ -2790,76 +2829,104 @@ SelectionDisplay = (function() {
...
+			
+			var localDeltaPivot = deltaPivot;
+			var localSigns = signs;
 
             var pickRay = generalComputePickRay(event.x, event.y);
-            newPick = rayPlaneIntersection(pickRay,
+			
+			// Are we using handControllers or Mouse - only relevant for 3D tools
+			var controllerPose = getControllerWorldLocation(activeHand, true);
+			if (HMD.isHMDAvailable() && HMD.isHandControllerAvailable() && controllerPose.valid && that.triggered && directionFor3DStretch ){
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: spaces

File: scripts/system/libraries/entitySelectionTool.js [hunk 7]
```diff
@@ -2790,76 +2829,104 @@ SelectionDisplay = (function() {
...
+				localDeltaPivot = deltaPivot3D;
+
+				newPick = pickRay.origin;
+			
+				var vector = Vec3.subtract(newPick, lastPick3D);
+				
+				vector = Vec3.multiplyQbyV(Quat.inverse(rotation), vector);
+			
+				// Range of Motion
+				if (distanceFor3DStretch > DISTANCE_INFLUENCE_THRESHOLD){
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: spaces

File: scripts/system/libraries/entitySelectionTool.js [hunk 8]
```diff
@@ -2790,76 +2829,104 @@ SelectionDisplay = (function() {
...
+					// Range of Motion
+					vector = Vec3.multiply(distanceFor3DStretch , vector);
+				}
+				
+				localSigns = directionFor3DStretch;
+				
+			}else{
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: spaces

**Comment by hifi-gustavo** [comment]
Build console output for commit: 19de579057b5847fca1444e40acbae2c7a722e76 available at: https://gist.github.com/7d76fd2ec8d71749684f42c203be06a4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-19de579.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-19de579.dmg

File: scripts/system/libraries/entitySelectionTool.js (hunk 1 above)
Comment (CraftsmanMG) [inline scripts/system/libraries/entitySelectionTool.js]: Thank you.

**Comment by CraftsmanMG** [comment]
Improved the style.

**Comment by hifi-gustavo** [comment]
Build console output for commit: fcde85e51b9f0e73870a26494f13945040f1b017 available at: https://gist.github.com/7d76fd2ec8d71749684f42c203be06a4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-fcde85e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-fcde85e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 70239b560a8d23ec09efa4c190e99ab8692b33c6 available at: https://gist.github.com/7d76fd2ec8d71749684f42c203be06a4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-70239b5.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-70239b5.dmg

**Comment by CraftsmanMG** [comment]
Testing Notes:
Prerequesites: HMD + motion controllers, permission to edit entities.

Open up Edit mode on tablet.
Create a Cube entity. Position your avatar in a range of 1 meter (just a little bit more than at arms reach).
Expected - A cube appears with grey tool overlays.

Point the laser at one of the bottom corner overlays and pull the trigger. Move your motion controller in various directions while holding trigger. Try the other overlays on the corners.
Expected - The vertex of the cube moves according to your hand controller motion in 3 dimensions.

Move the cube farther away from your avatar. Repeat last step.
Expected - The amount of resizing produced by the hand controller movement increases with the distance to the object.

Switch to mouse control while wearing the HMD and click and drag on the grey bottom corner overlays of the entity.
Expected - You are able to resize the entity on a 2D plane.

Switch again to motion controllers and select the bottom corner overlay and move your hand.
Expected - The vertex of the cube moves accordind to your hand controller motion in 3 dimensions.

**Comment by amvmoody** [comment]
Pass

File: scripts/system/libraries/entitySelectionTool.js [hunk 9]
```diff
@@ -2652,8 +2668,20 @@ SelectionDisplay = (function() {
...
             pickRayPosition = Vec3.sum(initialPosition, Vec3.multiplyQbyV(rotation, scaledOffsetWorld));
-
+            
+            if (directionFor3DStretch) {
+                // pivot, offset and pickPlanePosition for 3D manipulation
+                var scaledPivot3D = Vec3.multiply(0.5, Vec3.multiply(1.0, directionFor3DStretch));
+                deltaPivot3D = Vec3.subtract(centeredRP, scaledPivot3D);
+                
+                var scaledOffsetWorld3D = vec3Mult(initialDimensions, 
+                Vec3.subtract(Vec3.multiply(0.5, Vec3.multiply(-1.0, directionFor3DStretch)),
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: indent is wrong on this. 

> Multi-line statements must have all n+1 lines indented at least one level (four spaces).
> Align all n+2 lines with the indentation of the n+1 line.

File: scripts/system/libraries/entitySelectionTool.js [hunk 10]
```diff
@@ -2790,24 +2831,49 @@ SelectionDisplay = (function() {
...
+            
+            var localDeltaPivot = deltaPivot;
+            var localSigns = signs;
 
             var pickRay = generalComputePickRay(event.x, event.y);
-            newPick = rayPlaneIntersection(pickRay,
+            
+            // Are we using handControllers or Mouse - only relevant for 3D tools
+            var controllerPose = getControllerWorldLocation(activeHand, true);
+            if (HMD.isHMDAvailable() && HMD.isHandControllerAvailable() && controllerPose.valid && that.triggered && directionFor3DStretch ) {
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: coding standard: no space between `directionFor3DStretch` and `)`

**Comment by CraftsmanMG** [comment]
Well spotted, thank you @ZappoMan I just pushed. Could you confirm if the multiline is correct like this?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 956d1e9d95c84b7396edc464b055fcd34d4bd32d available at: https://gist.github.com/7d76fd2ec8d71749684f42c203be06a4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-956d1e9.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-956d1e9.dmg

File: scripts/system/libraries/entitySelectionTool.js [hunk 11]
```diff
@@ -2754,12 +2782,25 @@ SelectionDisplay = (function() {
...
-
+                
+            var planeNormal3D = {
+                x: 0,
+                y: 0,
+                z: 0
+            };
+            if (directionFor3DStretch) {
+                lastPick3D = rayPlaneIntersection(pickRay,
+                pickRayPosition3D,
```
Comment (ZappoMan) [inline scripts/system/libraries/entitySelectionTool.js]: this is still not formatted correctly... you need to indent lines 2799-2800 but at least 1 more indent (or you could leave them on line 1 if they are less than 128 columns wide.

Comment (CraftsmanMG) [inline scripts/system/libraries/entitySelectionTool.js]: Thank you for your patience, we're new to this coding standard. (Fix indent lines 2799-2800
Fix one line that had more than 128 columns wide)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 27029eb8db3e4b574de5c76e4bed3bef51e7adde available at: https://gist.github.com/7d76fd2ec8d71749684f42c203be06a4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-27029eb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10182-27029eb.dmg

**Comment by themelissabrown** [comment]
@ZappoMan is this ready to approve and merge now?

**Comment by PhilipRosedale** [comment]
Tried it out, and works as expected/described.  Thx!

**Pull request merged**